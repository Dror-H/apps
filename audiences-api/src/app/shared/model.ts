export interface SynonymsResponse {
  text: string;
  sense: string;
  results: Result[];
}

export interface Result {
  score: number;
  text: string;
  count: number;
}
export interface SearchResult<Spec = FacebookTargeting> {
  id: string;
  spec: Spec;
  topics: string[];
  userTags?: string[];
  rank: number;
  specRatio?: { [k: string]: number };
  createdByContributors?: boolean;
}

export interface FacebookTargeting {
  id: string;
  name: string;
  campaign?: Campaign;
  targeting: Targeting;
}

export interface Campaign {
  id: string;
  name: string;
}

export interface Segment {
  id: string;
  name: string;
  audience_size_lower_bound: number;
  audience_size_upper_bound: number;
  path: string[];
  size?: number;
  description: string;
  topic: string;
  type?: keyof FlexibleSpec | string;
  disambiguation_category?: string;
  valid?: boolean;
  relatedWords?: string[];
  status?: string;
  synonyms?: string[];
  datamuse?: string[];
}

export interface Targeting {
  age_max?: number;
  age_min?: number;
  locales?: number[];
  flexible_spec?: FlexibleSpec[];
  exclusions?: FlexibleSpec;
  geo_locations?: GeoLocations;
  device_platforms?: string[];
  facebook_positions?: string[];
  instagram_positions?: string[];
  publisher_platforms?: string[];
  targeting_optimization?: string;
}

export interface FlexibleSpec {
  relationship_statuses?: number[];
  college_years?: number[];
  education_statuses?: number[];
  demographics?: Segment[];
  connections?: Segment[];
  friends_of_connections?: Segment[];
  custom_audiences?: Segment[];
  interests?: Segment[];
  behaviors?: Segment[];
  education_majors?: Segment[];
  education_schools?: Segment[];
  family_statuses?: Segment[];
  home_value?: Segment[];
  interested_in?: Segment[];
  income?: Segment[];
  industries?: Segment[];
  life_events?: Segment[];
  user_adclusters?: Segment[];
  work_positions?: Segment[];
  work_employers?: Segment[];
}

export const GeoLocationsObjectKeys = ['cities', 'neighborhoods', 'places', 'regions'];
export const GeoLocationsStringKeys = ['countries', 'location_types'];
export interface GeoLocations {
  cities: City[];
  neighborhoods: Neighborhood[];
  places: Place[];
  regions: Region[];
  countries: string[];
  location_types: string[];
}
export interface Neighborhood {
  key: string;
  name: string;
  radius: number;
  region: string;
  country: string;
  region_id: string;
}

export interface City {
  key: string;
  name: string;
  radius: number;
  region: string;
  country: string;
  region_id: string;
  distance_unit: string;
}

export interface Place {
  key: string;
  name: string;
  radius: number;
  country: string;
  latitude: number;
  longitude: number;
  region_id: number;
  distance_unit: string;
  primary_city_id: number;
}

export interface Region {
  key: string;
  name: string;
  country: string;
}

export interface LdaTopic {
  term: string;
  probability: number;
}
