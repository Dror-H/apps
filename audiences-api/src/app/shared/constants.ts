export const CATEGORY_ID_PREFIX = {
  education_statuses: 100,
  relationship_statuses: 200,
  user_device: 300,
  user_os: 400,
};
