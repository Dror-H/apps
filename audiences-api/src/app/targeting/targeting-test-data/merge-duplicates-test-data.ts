import { SearchResult, FlexibleSpec, Targeting } from './../../shared/model';

export const flexSpecsNullSegments: FlexibleSpec[] = [
  {
    interests: [
      {
        id: '1',
        name: '',
        audience_size_lower_bound: 0,
        audience_size_upper_bound: 0,
        path: ['test'],
        description: 'double location = exclusion.interests[1]',
        topic: 'test',
      },
    ],
    family_statuses: [null],
  },
  {
    interests: [null],
    family_statuses: [
      {
        id: '1',
        name: '',
        audience_size_lower_bound: 0,
        audience_size_upper_bound: 0,
        path: ['test'],
        description: 'double location = exclusion.interests[1]',
        topic: 'test',
      },
    ],
  },
];

export const targetingNullSegments: Targeting = {
  exclusions: {
    interests: [
      {
        id: '1',
        name: '',
        audience_size_lower_bound: 0,
        audience_size_upper_bound: 0,
        path: ['test'],
        description: 'double location = flexible_spec.interests[0]',
        topic: 'test',
      },
      {
        id: '2',
        name: '',
        audience_size_lower_bound: 0,
        audience_size_upper_bound: 0,
        path: ['test'],
        description: 'double location = exclusion.interests[2]',
        topic: 'test',
      },
      null,
    ],
    family_statuses: [null],
  },
  flexible_spec: [
    {
      interests: [
        {
          id: '1',
          name: '',
          audience_size_lower_bound: 0,
          audience_size_upper_bound: 0,
          path: ['test'],
          description: 'double location = exclusion.interests[1]',
          topic: 'test',
        },
      ],
      family_statuses: [null],
    },
    {
      interests: [
        {
          id: '2',
          name: '',
          audience_size_lower_bound: 0,
          audience_size_upper_bound: 0,
          path: ['test'],
          description: 'double location = exclusion.interests[1]',
          topic: 'test',
        },
      ],
      work_employers: [null],
    },
  ],
};

export const mockIncludesSegmentIfInExclusion: SearchResult = {
  id: '38fe6309-9085-5711-8bf7-a80f7cf38877',
  createdByContributors: false,
  spec: {
    id: '154728e7-d715-47d9-aabd-8b97903a7362',
    name: '',
    targeting: {
      age_min: 18,
      age_max: 65,
      locales: [],
      exclusions: {
        interests: [
          {
            id: '1',
            name: '',
            audience_size_lower_bound: 0,
            audience_size_upper_bound: 0,
            path: ['test'],
            description: 'double location = flexible_spec.interests[0]',
            topic: 'test',
          },
          {
            id: '2',
            name: '',
            audience_size_lower_bound: 0,
            audience_size_upper_bound: 0,
            path: ['test'],
            description: 'double location = exclusion.interests[2]',
            topic: 'test',
          },
        ],
      },
      flexible_spec: [
        {
          interests: [
            {
              id: '1',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double location = exclusion.interests[1]',
              topic: 'test',
            },
          ],
        },
        {
          interests: [
            {
              id: '2',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double location = exclusion.interests[1]',
              topic: 'test',
            },
          ],
        },
      ],
      geo_locations: {
        countries: [],
        location_types: [],
        cities: [],
        neighborhoods: [],
        places: [],
        regions: [],
      },
      targeting_optimization: 'none',
    },
  },
  userTags: [],
  topics: [],
  rank: 100,
};

export const mockDeduplicateIncludes: SearchResult = {
  id: '38fe6309-9085-5711-8bf7-a80f7cf38877',
  createdByContributors: false,
  spec: {
    id: '154728e7-d715-47d9-aabd-8b97903a7362',
    name: '',
    targeting: {
      age_min: 18,
      age_max: 65,
      locales: [],
      exclusions: {},
      flexible_spec: [
        {
          interests: [
            {
              id: '1',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double locations = [flexible_spec[2]interests[0], flexible_spec[2]interests[1]]',
              topic: 'test',
            },
            {
              id: '1',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double to be removed',
              topic: 'test',
            },
            {
              id: '2',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'test',
              topic: 'test',
            },
          ],
          work_employers: [
            {
              id: '1',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double to be removed',
              topic: 'test',
            },
            {
              id: '3',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'test',
              topic: 'test',
            },
          ],
        },
        {
          interests: [
            {
              id: '1',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double to be removed',
              topic: 'test',
            },
          ],
          work_employers: [
            {
              id: '1',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double to be removed',
              topic: 'test',
            },
            {
              id: '4',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'test',
              topic: 'test',
            },
          ],
          work_positions: [
            {
              id: '1',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double to be removed',
              topic: 'test',
            },
          ],
        },
        {
          interests: [
            {
              id: '1',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double to be removed',
              topic: 'test',
            },
          ],
          work_employers: [
            {
              id: '1',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double to be removed',
              topic: 'test',
            },
          ],
          education_statuses: [1, 2, 3],
          work_positions: [
            {
              id: '1',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double to be removed',
              topic: 'test',
            },
          ],
        },
      ],
      geo_locations: {
        countries: [],
        location_types: [],
        cities: [],
        neighborhoods: [],
        places: [],
        regions: [],
      },
      targeting_optimization: 'none',
    },
  },
  userTags: [],
  topics: [],
  rank: 100,
};

export const mockDeduplicateAll: SearchResult = {
  id: '38fe6309-9085-5711-8bf7-a80f7cf38877',
  createdByContributors: false,
  spec: {
    id: '154728e7-d715-47d9-aabd-8b97903a7362',
    name: '',
    targeting: {
      age_min: 18,
      age_max: 65,
      locales: [],
      exclusions: {
        interests: [
          {
            id: '1',
            name: '',
            audience_size_lower_bound: 0,
            audience_size_upper_bound: 0,
            path: ['test'],
            description: 'double location = flexible_spec.interests[0]',
            topic: 'test',
          },
          {
            id: '2',
            name: '',
            audience_size_lower_bound: 0,
            audience_size_upper_bound: 0,
            path: ['test'],
            description: 'double location = exclusion.interests[2]',
            topic: 'test',
          },
        ],
      },
      flexible_spec: [
        {
          interests: [
            {
              id: '1',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double locations = [flexible_spec[2]interests[0], flexible_spec[2]interests[1]]',
              topic: 'test',
            },
            {
              id: '1',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double to be removed',
              topic: 'test',
            },
            {
              id: '2',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'test',
              topic: 'test',
            },
          ],
          work_employers: [
            {
              id: '1',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double to be removed',
              topic: 'test',
            },
            {
              id: '3',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'test',
              topic: 'test',
            },
          ],
        },
        {
          interests: [
            {
              id: '1',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double to be removed',
              topic: 'test',
            },
          ],
          work_employers: [
            {
              id: '1',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double to be removed',
              topic: 'test',
            },
            {
              id: '4',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'test',
              topic: 'test',
            },
          ],
          work_positions: [
            {
              id: '1',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double to be removed',
              topic: 'test',
            },
          ],
        },
        {
          interests: [
            {
              id: '1',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double to be removed',
              topic: 'test',
            },
          ],
          work_employers: [
            {
              id: '1',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double to be removed',
              topic: 'test',
            },
          ],
          education_statuses: [1, 2, 3],
          work_positions: [
            {
              id: '1',
              name: '',
              audience_size_lower_bound: 0,
              audience_size_upper_bound: 0,
              path: ['test'],
              description: 'double to be removed',
              topic: 'test',
            },
          ],
        },
      ],
      geo_locations: {
        countries: [],
        location_types: [],
        cities: [],
        neighborhoods: [],
        places: [],
        regions: [],
      },
      targeting_optimization: 'none',
    },
  },
  userTags: [],
  topics: [],
  rank: 100,
};
