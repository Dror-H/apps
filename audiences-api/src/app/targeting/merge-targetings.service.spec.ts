import { Test } from '@nestjs/testing';
import { GeoLocations } from '../shared/model';
import {
  flexSpecsNullSegments,
  mockDeduplicateAll,
  mockDeduplicateIncludes,
  mockIncludesSegmentIfInExclusion,
  targetingNullSegments,
} from '../targeting/targeting-test-data/merge-duplicates-test-data';
import { MergeTargetingsService } from './merge-targetings.service';
import { AudiencesToMerge, mockMergedSingleAudience } from './targeting-test-data/merge-test-data';
const geoLocations = [
  '{"countries": ["US"], "location_types": ["home", "recent"]}',
  '{"regions": [{"key": "4079", "name": "England", "country": "GB"}, {"key": "4081", "name": "Scotland", "country": "GB"}], "countries": ["US", "IE", "CA", "AU"], "location_types": ["home", "recent"]}',
  '{"zips": [{"key": "US:79938", "name": "79938", "country": "US", "region_id": 3886, "primary_city_id": 2525963}], "location_types": ["home", "recent"], "custom_locations": [{"name": "11140 Montana Ave, El Paso, Texas, United States", "radius": 10, "country": "US", "latitude": 31.80031, "longitude": -106.31022, "region_id": 3886, "distance_unit": "mile", "address_string": "11140 Montana Ave, El Paso, Texas, United States", "primary_city_id": 2526788}]}',
  `{"cities": [{"key": "171230", "name": "Antwerp", "radius": 10, "region": "Flemish Region", "country": "BE", "region_id": "4136", "distance_unit": "mile"}, {"key": "2367397", "name": "Dnipro", "radius": 10, "region": "Dnipropetrovsk Oblast", "country": "UA", "region_id": "3781", "distance_unit": "mile"}, {"key": "2384095", "name": "Odessa", "radius": 10, "region": "Odessa Oblast", "country": "UA", "region_id": "3794", "distance_unit": "mile"}, {"key": "313044", "name": "Z\\u00fcrich", "radius": 10, "region": "Z\\u00fcrich", "country": "CH", "region_id": "603", "distance_unit": "mile"}, {"key": "542609", "name": "Berlin", "region": "Berlin", "country": "DE", "region_id": "840", "distance_unit": "mile"}, {"key": "550095", "name": "D\\u00fcsseldorf", "radius": 10, "region": "Nordrhein-Westfalen", "country": "DE", "region_id": "831", "distance_unit": "mile"}, {"key": "554021", "name": "Frankfurt", "radius": 10, "region": "Hessen", "country": "DE", "region_id": "829", "distance_unit": "mile"}, {"key": "579270", "name": "Munich", "radius": 10, "region": "Bayern", "country": "DE", "region_id": "826", "distance_unit": "mile"}, {"key": "771540", "name": "Marseille", "radius": 10, "region": "Provence-Alpes-C\\u00f4te d'Azur", "country": "FR", "region_id": "1087", "distance_unit": "mile"}, {"key": "777934", "name": "Paris", "radius": 10, "region": "\\u00cele-de-France", "country": "FR", "region_id": "1078", "distance_unit": "mile"}, {"key": "812057", "name": "London", "radius": 10, "region": "England", "country": "GB", "region_id": "4079", "distance_unit": "mile"}], "regions": [{"key": "1625", "name": "Budapest", "country": "HU"}, {"key": "3146", "name": "Moscow", "country": "RU"}], "location_types": ["home"], "custom_locations": [{"radius": 5, "country": "US", "latitude": 40.676993, "longitude": -73.901595, "region_id": 3875, "distance_unit": "mile", "primary_city_id": 2490299}]}`,
  '{"cities": [{"key": "2420379", "name": "Los Angeles", "region": "California", "country": "US", "region_id": "3847", "distance_unit": "mile"}, {"key": "2421836", "name": "San Francisco", "region": "California", "country": "US", "region_id": "3847", "distance_unit": "mile"}, {"key": "2490299", "name": "New York", "region": "New York", "country": "US", "region_id": "3875", "distance_unit": "mile"}], "places": [{"key": "155403677865323", "name": "BAY AREA", "radius": 2, "country": "US", "latitude": 38.123625, "longitude": -122.217148, "region_id": 3847, "distance_unit": "kilometer", "primary_city_id": 2422640}], "regions": [{"key": "3849", "name": "Connecticut", "country": "US"}], "location_types": ["home", "recent"], "medium_geo_areas": [{"key": "2790231", "name": "Orange County", "region": "California", "country": "US", "region_id": "3847"}]}',
  '{"countries": ["US", "IE", "IT", "NL", "NZ", "NO", "CA", "ES", "CH", "GB", "DK", "BE", "AU", "AT", "PT", "IS", "FR", "DE"], "location_types": ["home", "recent"]}',
  '{"cities": [{"key": "803904", "name": "Birmingham", "radius": 50, "region": "England", "country": "GB", "region_id": "4079", "distance_unit": "mile"}], "location_types": ["home"]}',
  '{"countries": ["US", "CA"], "location_types": ["home", "recent"]}',
  '{"countries": ["US"], "location_types": ["home"]}',
  '{"countries": ["US", "IE", "IT", "MC", "NL", "CA", "ES", "CH", "GB", "BE", "AT", "PT", "LU", "FR", "DE"], "location_types": ["home"]}',
  '{"regions": [{"key": "3849", "name": "Connecticut", "country": "US"}, {"key": "3862", "name": "Maine", "country": "US"}, {"key": "3864", "name": "Massachusetts", "country": "US"}, {"key": "3872", "name": "New Hampshire", "country": "US"}, {"key": "3875", "name": "New York", "country": "US"}, {"key": "3882", "name": "Rhode Island", "country": "US"}, {"key": "3888", "name": "Vermont", "country": "US"}], "location_types": ["home", "recent"]}',
  '{"countries": ["US", "IE", "IT", "NL", "NZ", "NO", "CA", "ES", "CH", "GB", "DK", "BE", "AU", "AT", "PT", "IS", "FR", "DE"], "location_types": ["home", "recent"]}',
  '{"countries": ["CH"], "location_types": ["home", "recent"]}',
  '{"countries": ["GB"], "location_types": ["home", "recent"]}',
  'null',
];

const flexibleSpecs = [
  [{ interests: [] }],
  [
    {
      interests: [{ id: '6002926675372', name: 'Scholarship' }],
      education_schools: [{ id: '475750562498596', name: 'Study in UK' }],
    },
  ],
  [
    {
      interests: [
        { id: '6002964301046', name: 'Business school' },
        { id: '6002926675372', name: 'Scholarship' },
      ],
      college_years: [2022, 2023, 2024, 2025, 2026],
      education_majors: [{ id: '106335129403847', name: 'Business management' }],
      education_statuses: [4, 5],
    },
    {
      interests: [{ id: '6003748928462', name: 'Destroyer' }],
    },
  ],
  [
    {
      interests: [
        { id: '6002926675372', name: 'Scholarship' },
        { id: '6003057317756', name: 'Business education' },
      ],
      education_statuses: [2, 100],
    },
    {
      interests: [{ id: '6004000198906', name: 'Professional development' }],
    },
  ],
] as any;

describe('MergeTargetings Test suite', () => {
  let service: MergeTargetingsService;

  beforeEach(async () => {
    const module = await Test.createTestingModule({
      providers: [MergeTargetingsService],
    }).compile();

    service = module.get<MergeTargetingsService>(MergeTargetingsService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should get max narrow', () => {
    const max = service.maxNarrowSpec(flexibleSpecs);
    expect(max).toEqual(2);
  });

  it('should merge specs', () => {
    const merged = service.mergeFlexibleSpec(flexibleSpecs);
    expect(merged).toMatchSnapshot();
  });

  it('should merge locations', () => {
    const locations = service.mergeGeoLocations(geoLocations.map((v) => JSON.parse(v)) as GeoLocations[]);
    expect(locations).toMatchSnapshot();
  });

  describe('merge', () => {
    it('should return same search result as provided', () => {
      const mergedAudience = service.merge({ audiences: [mockMergedSingleAudience] });
      expect(mergedAudience).toEqual(mergedAudience);
    });

    it('should merge', () => {
      // TODO - fix Segment type
      const mergedAudience = service.merge({ audiences: AudiencesToMerge as any });
      expect(mergedAudience).toEqual(mergedAudience);
    });
  });

  describe('deduplicateIncludesAndExclusion', () => {
    it('should remove all duplicates', () => {
      service.deduplicateIncludesAndExclusion(
        mockDeduplicateAll.spec.targeting.flexible_spec,
        mockDeduplicateAll.spec.targeting.exclusions,
      );
      expect(mockDeduplicateAll.spec.targeting).toMatchSnapshot();
    });

    it('should remove duplicates and null values from exclusion', () => {
      service.deduplicateIncludesAndExclusion(targetingNullSegments.flexible_spec, targetingNullSegments.exclusions);
      expect(targetingNullSegments).toMatchSnapshot();
    });

    it('should remove duplicates and number type segments', () => {
      service.deduplicateIncludes(
        mockDeduplicateIncludes.spec.targeting.flexible_spec,
        mockDeduplicateIncludes.spec.targeting.flexible_spec[0].interests[0],
      );
      expect(mockDeduplicateIncludes.spec.targeting.flexible_spec).toMatchSnapshot();
    });

    it('should remove duplicates and null values', () => {
      service.deduplicateIncludes(flexSpecsNullSegments, flexSpecsNullSegments[0].interests[0]);
      expect(flexSpecsNullSegments).toMatchSnapshot();
    });
  });

  describe('removeIncludesSegmentIfInExclusion', () => {
    it('should remove segment from exclusion if same segment id exists in inclusion', () => {
      [0, 1].forEach((index) => {
        service.removeIncludesSegmentIfInExclusion(
          mockIncludesSegmentIfInExclusion.spec.targeting.exclusions,
          mockIncludesSegmentIfInExclusion.spec.targeting.flexible_spec[index].interests[0],
          'interests',
        );
      });
      expect(mockIncludesSegmentIfInExclusion.spec.targeting.exclusions.interests).toEqual([]);
    });
  });
});
