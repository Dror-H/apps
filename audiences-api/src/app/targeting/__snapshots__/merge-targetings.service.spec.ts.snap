// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MergeTargetings Test suite deduplicateIncludesAndExclusion should remove all duplicates 1`] = `
Object {
  "age_max": 65,
  "age_min": 18,
  "exclusions": Object {
    "interests": Array [],
  },
  "flexible_spec": Array [
    Object {
      "interests": Array [
        Object {
          "audience_size_lower_bound": 0,
          "audience_size_upper_bound": 0,
          "description": "double locations = [flexible_spec[2]interests[0], flexible_spec[2]interests[1]]",
          "id": "1",
          "name": "",
          "path": Array [
            "test",
          ],
          "topic": "test",
        },
        Object {
          "audience_size_lower_bound": 0,
          "audience_size_upper_bound": 0,
          "description": "test",
          "id": "2",
          "name": "",
          "path": Array [
            "test",
          ],
          "topic": "test",
        },
      ],
      "work_employers": Array [
        Object {
          "audience_size_lower_bound": 0,
          "audience_size_upper_bound": 0,
          "description": "test",
          "id": "3",
          "name": "",
          "path": Array [
            "test",
          ],
          "topic": "test",
        },
      ],
    },
    Object {
      "interests": Array [],
      "work_employers": Array [
        Object {
          "audience_size_lower_bound": 0,
          "audience_size_upper_bound": 0,
          "description": "test",
          "id": "4",
          "name": "",
          "path": Array [
            "test",
          ],
          "topic": "test",
        },
      ],
      "work_positions": Array [],
    },
    Object {
      "education_statuses": Array [],
      "interests": Array [],
      "work_employers": Array [],
      "work_positions": Array [],
    },
  ],
  "geo_locations": Object {
    "cities": Array [],
    "countries": Array [],
    "location_types": Array [],
    "neighborhoods": Array [],
    "places": Array [],
    "regions": Array [],
  },
  "locales": Array [],
  "targeting_optimization": "none",
}
`;

exports[`MergeTargetings Test suite deduplicateIncludesAndExclusion should remove duplicates and null values 1`] = `
Array [
  Object {
    "family_statuses": Array [],
    "interests": Array [
      Object {
        "audience_size_lower_bound": 0,
        "audience_size_upper_bound": 0,
        "description": "double location = exclusion.interests[1]",
        "id": "1",
        "name": "",
        "path": Array [
          "test",
        ],
        "topic": "test",
      },
    ],
  },
  Object {
    "family_statuses": Array [],
    "interests": Array [],
  },
]
`;

exports[`MergeTargetings Test suite deduplicateIncludesAndExclusion should remove duplicates and null values from exclusion 1`] = `
Object {
  "exclusions": Object {
    "family_statuses": Array [],
    "interests": Array [],
  },
  "flexible_spec": Array [
    Object {
      "family_statuses": Array [],
      "interests": Array [
        Object {
          "audience_size_lower_bound": 0,
          "audience_size_upper_bound": 0,
          "description": "double location = exclusion.interests[1]",
          "id": "1",
          "name": "",
          "path": Array [
            "test",
          ],
          "topic": "test",
        },
      ],
    },
    Object {
      "interests": Array [
        Object {
          "audience_size_lower_bound": 0,
          "audience_size_upper_bound": 0,
          "description": "double location = exclusion.interests[1]",
          "id": "2",
          "name": "",
          "path": Array [
            "test",
          ],
          "topic": "test",
        },
      ],
      "work_employers": Array [],
    },
  ],
}
`;

exports[`MergeTargetings Test suite deduplicateIncludesAndExclusion should remove duplicates and number type segments 1`] = `
Array [
  Object {
    "interests": Array [
      Object {
        "audience_size_lower_bound": 0,
        "audience_size_upper_bound": 0,
        "description": "double locations = [flexible_spec[2]interests[0], flexible_spec[2]interests[1]]",
        "id": "1",
        "name": "",
        "path": Array [
          "test",
        ],
        "topic": "test",
      },
      Object {
        "audience_size_lower_bound": 0,
        "audience_size_upper_bound": 0,
        "description": "test",
        "id": "2",
        "name": "",
        "path": Array [
          "test",
        ],
        "topic": "test",
      },
    ],
    "work_employers": Array [
      Object {
        "audience_size_lower_bound": 0,
        "audience_size_upper_bound": 0,
        "description": "test",
        "id": "3",
        "name": "",
        "path": Array [
          "test",
        ],
        "topic": "test",
      },
    ],
  },
  Object {
    "interests": Array [],
    "work_employers": Array [
      Object {
        "audience_size_lower_bound": 0,
        "audience_size_upper_bound": 0,
        "description": "test",
        "id": "4",
        "name": "",
        "path": Array [
          "test",
        ],
        "topic": "test",
      },
    ],
    "work_positions": Array [],
  },
  Object {
    "education_statuses": Array [],
    "interests": Array [],
    "work_employers": Array [],
    "work_positions": Array [],
  },
]
`;

exports[`MergeTargetings Test suite should merge locations 1`] = `
Object {
  "cities": Array [
    Object {
      "country": "BE",
      "distance_unit": "mile",
      "key": "171230",
      "name": "Antwerp",
      "radius": 10,
      "region": "Flemish Region",
      "region_id": "4136",
    },
    Object {
      "country": "UA",
      "distance_unit": "mile",
      "key": "2367397",
      "name": "Dnipro",
      "radius": 10,
      "region": "Dnipropetrovsk Oblast",
      "region_id": "3781",
    },
    Object {
      "country": "UA",
      "distance_unit": "mile",
      "key": "2384095",
      "name": "Odessa",
      "radius": 10,
      "region": "Odessa Oblast",
      "region_id": "3794",
    },
    Object {
      "country": "CH",
      "distance_unit": "mile",
      "key": "313044",
      "name": "Zürich",
      "radius": 10,
      "region": "Zürich",
      "region_id": "603",
    },
    Object {
      "country": "DE",
      "distance_unit": "mile",
      "key": "542609",
      "name": "Berlin",
      "region": "Berlin",
      "region_id": "840",
    },
    Object {
      "country": "DE",
      "distance_unit": "mile",
      "key": "550095",
      "name": "Düsseldorf",
      "radius": 10,
      "region": "Nordrhein-Westfalen",
      "region_id": "831",
    },
    Object {
      "country": "DE",
      "distance_unit": "mile",
      "key": "554021",
      "name": "Frankfurt",
      "radius": 10,
      "region": "Hessen",
      "region_id": "829",
    },
    Object {
      "country": "DE",
      "distance_unit": "mile",
      "key": "579270",
      "name": "Munich",
      "radius": 10,
      "region": "Bayern",
      "region_id": "826",
    },
    Object {
      "country": "FR",
      "distance_unit": "mile",
      "key": "771540",
      "name": "Marseille",
      "radius": 10,
      "region": "Provence-Alpes-Côte d'Azur",
      "region_id": "1087",
    },
    Object {
      "country": "FR",
      "distance_unit": "mile",
      "key": "777934",
      "name": "Paris",
      "radius": 10,
      "region": "Île-de-France",
      "region_id": "1078",
    },
    Object {
      "country": "GB",
      "distance_unit": "mile",
      "key": "812057",
      "name": "London",
      "radius": 10,
      "region": "England",
      "region_id": "4079",
    },
    Object {
      "country": "US",
      "distance_unit": "mile",
      "key": "2420379",
      "name": "Los Angeles",
      "region": "California",
      "region_id": "3847",
    },
    Object {
      "country": "US",
      "distance_unit": "mile",
      "key": "2421836",
      "name": "San Francisco",
      "region": "California",
      "region_id": "3847",
    },
    Object {
      "country": "US",
      "distance_unit": "mile",
      "key": "2490299",
      "name": "New York",
      "region": "New York",
      "region_id": "3875",
    },
    Object {
      "country": "GB",
      "distance_unit": "mile",
      "key": "803904",
      "name": "Birmingham",
      "radius": 50,
      "region": "England",
      "region_id": "4079",
    },
  ],
  "countries": Array [
    "US",
    "IE",
    "CA",
    "AU",
    "IT",
    "NL",
    "NZ",
    "NO",
    "ES",
    "CH",
    "GB",
    "DK",
    "BE",
    "AT",
    "PT",
    "IS",
    "FR",
    "DE",
    "MC",
    "LU",
  ],
  "location_types": Array [
    "home",
    "recent",
  ],
  "places": Array [
    Object {
      "country": "US",
      "distance_unit": "kilometer",
      "key": "155403677865323",
      "latitude": 38.123625,
      "longitude": -122.217148,
      "name": "BAY AREA",
      "primary_city_id": 2422640,
      "radius": 2,
      "region_id": 3847,
    },
  ],
  "regions": Array [
    Object {
      "country": "GB",
      "key": "4079",
      "name": "England",
    },
    Object {
      "country": "GB",
      "key": "4081",
      "name": "Scotland",
    },
    Object {
      "country": "HU",
      "key": "1625",
      "name": "Budapest",
    },
    Object {
      "country": "RU",
      "key": "3146",
      "name": "Moscow",
    },
    Object {
      "country": "US",
      "key": "3849",
      "name": "Connecticut",
    },
    Object {
      "country": "US",
      "key": "3862",
      "name": "Maine",
    },
    Object {
      "country": "US",
      "key": "3864",
      "name": "Massachusetts",
    },
    Object {
      "country": "US",
      "key": "3872",
      "name": "New Hampshire",
    },
    Object {
      "country": "US",
      "key": "3875",
      "name": "New York",
    },
    Object {
      "country": "US",
      "key": "3882",
      "name": "Rhode Island",
    },
    Object {
      "country": "US",
      "key": "3888",
      "name": "Vermont",
    },
  ],
}
`;

exports[`MergeTargetings Test suite should merge specs 1`] = `
Array [
  Object {
    "college_years": Array [
      2022,
      2023,
      2024,
      2025,
      2026,
    ],
    "education_majors": Array [
      Object {
        "id": "106335129403847",
        "name": "Business management",
      },
    ],
    "education_schools": Array [
      Object {
        "id": "475750562498596",
        "name": "Study in UK",
      },
    ],
    "education_statuses": Array [
      4,
      5,
      2,
      100,
    ],
    "interests": Array [
      Object {
        "id": "6002926675372",
        "name": "Scholarship",
      },
      Object {
        "id": "6002964301046",
        "name": "Business school",
      },
      Object {
        "id": "6003057317756",
        "name": "Business education",
      },
    ],
  },
  Object {
    "interests": Array [
      Object {
        "id": "6003748928462",
        "name": "Destroyer",
      },
      Object {
        "id": "6004000198906",
        "name": "Professional development",
      },
    ],
  },
]
`;
